File: internal/api/handlers_tasks.go
package api

import (
	"log/slog"
	"net/http"
	"strconv"

	"github.com/go-chi/chi/v5"

	"github.com/Z-M-Huang/openhive/internal/domain"
)

const (
	defaultTasksLimit = 50
	maxTasksLimit     = 500
)

// tasksResponse wraps a paginated list of tasks.
type tasksResponse struct {
	Tasks   []*domain.Task `json:"tasks"`
	Total   int            `json:"total"`
	HasMore bool           `json:"has_more"`
	Limit   int            `json:"limit"`
	Offset  int            `json:"offset"`
}

// taskWithSubtree adds a subtasks array to a task for tree display.
type taskWithSubtree struct {
	*domain.Task
	Subtasks []*domain.Task `json:"subtasks,omitempty"`
}

// GetTasksHandler returns a handler for GET /api/v1/tasks.
// Supports query params: status, team, limit, offset.
func GetTasksHandler(taskStore domain.TaskStore, logger *slog.Logger) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		q := r.URL.Query()

		limit := defaultTasksLimit
		if limitStr := q.Get("limit"); limitStr != "" {
			l, err := strconv.Atoi(limitStr)
			if err != nil || l < 1 {
				Error(w, http.StatusBadRequest, "INVALID_PARAM", "limit must be a positive integer")
				return
			}
			if l > maxTasksLimit {
				l = maxTasksLimit
			}
			limit = l
		}

		offset := 0
		if offsetStr := q.Get("offset"); offsetStr != "" {
			o, err := strconv.Atoi(offsetStr)
			if err != nil || o < 0 {
				Error(w, http.StatusBadRequest, "INVALID_PARAM", "offset must be a non-negative integer")
				return
			}
			offset = o
		}

		var tasks []*domain.Task
		var err error

		teamSlug := q.Get("team")
		statusStr := q.Get("status")

		switch {
		case teamSlug != "":
			tasks, err = taskStore.ListByTeam(r.Context(), teamSlug)
		case statusStr != "":
			status, parseErr := domain.ParseTaskStatus(statusStr)
			if parseErr != nil {
				Error(w, http.StatusBadRequest, "INVALID_PARAM", "invalid task status")
				return
			}
			tasks, err = taskStore.ListByStatus(r.Context(), status)
		default:
			// No filter: list running tasks as a practical default.
			tasks, err = taskStore.ListByStatus(r.Context(), domain.TaskStatusRunning)
			if err != nil {
				// Fallback to empty list on error
				tasks = []*domain.Task{}
				err = nil
			}
		}

		if err != nil {
			logger.Error("failed to list tasks", "error", err)
			Error(w, http.StatusInternalServerError, "INTERNAL_ERROR", "failed to list tasks")
			return
		}

		total := len(tasks)
		hasMore := false

		// Apply pagination
		if offset < len(tasks) {
			end := offset + limit
			if end > len(tasks) {
				end = len(tasks)
			} else {
				hasMore = true
			}
			tasks = tasks[offset:end]
		} else {
			tasks = []*domain.Task{}
		}

		JSON(w, http.StatusOK, tasksResponse{
			Tasks:   tasks,
			Total:   total,
			HasMore: hasMore,
			Limit:   limit,
			Offset:  offset,
		})
	}
}

// GetTaskHandler returns a handler for GET /api/v1/tasks/{id}.
func GetTaskHandler(taskStore domain.TaskStore, logger *slog.Logger) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id := chi.URLParam(r, "id")
		if id == "" {
			Error(w, http.StatusBadRequest, "INVALID_PARAM", "id is required")
			return
		}

		task, err := taskStore.Get(r.Context(), id)
		if err != nil {
			MapDomainError(w, err)
			return
		}

		// Fetch subtree
		subtree, err := taskStore.GetSubtree(r.Context(), id)
		if err != nil {
			logger.Error("failed to get task subtree", "id", id, "error", err)
			// Non-fatal: return task without subtree
			JSON(w, http.StatusOK, taskWithSubtree{Task: task})
			return
		}

		// Filter out the root task itself from the subtree
		var subtasks []*domain.Task
		for _, t := range subtree {
			if t.ID != id {
				subtasks = append(subtasks, t)
			}
		}

		JSON(w, http.StatusOK, taskWithSubtree{Task: task, Subtasks: subtasks})
	}
}

// CancelTaskHandler returns a handler for POST /api/v1/tasks/{id}/cancel.
func CancelTaskHandler(orch domain.GoOrchestrator, taskStore domain.TaskStore, logger *slog.Logger) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		// Enforce JSON body (even though we don't read it) for CSRF protection
		ct := r.Header.Get("Content-Type")
		if ct != "application/json" {
			Error(w, http.StatusUnsupportedMediaType, "INVALID_CONTENT_TYPE",
				"Content-Type must be application/json")
			return
		}

		id := chi.URLParam(r, "id")
		if id == "" {
			Error(w, http.StatusBadRequest, "INVALID_PARAM", "id is required")
			return
		}

		if err := orch.CancelTask(r.Context(), id); err != nil {
			logger.Error("failed to cancel task", "id", id, "error", err)
			MapDomainError(w, err)
			return
		}

		// Return updated task
		task, err := taskStore.Get(r.Context(), id)
		if err != nil {
			MapDomainError(w, err)
			return
		}

		JSON(w, http.StatusOK, task)
	}
}

File: backend/src/api/handlers-tasks.ts
/**
 * OpenHive Backend - Task API Handlers
 *
 * Implements GET /api/v1/tasks (with status/team/pagination),
 * GET /api/v1/tasks/:id (with subtree), POST /api/v1/tasks/:id/cancel.
 * Fastify schema validation guards all routes; task ID pattern matches UUID format.
 *
 */

import type { FastifyInstance, FastifyReply, FastifyRequest } from 'fastify';

import type { GoOrchestrator, TaskStore } from '../domain/interfaces.js';
import type { TaskStatus } from '../domain/enums.js';
import type { Task } from '../domain/types.js';
import type { MiddlewareLogger } from './middleware.js';
import { mapDomainError, sendError, sendJSON } from './response.js';
import type { FastifyReplyShim } from './response.js';

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const DEFAULT_TASKS_LIMIT = 50;
const MAX_TASKS_LIMIT = 500;

// ---------------------------------------------------------------------------
// Response types
// ---------------------------------------------------------------------------

/** Paginated task list response. */
export interface TasksResponse {
  tasks: Task[];
  total: number;
  has_more: boolean;
  limit: number;
  offset: number;
}

/** Task with optional subtasks array for tree display. */
export type TaskWithSubtree = Task & { subtasks?: Task[] };

// ---------------------------------------------------------------------------
// Internal request types
// ---------------------------------------------------------------------------

interface GetTasksQuery {
  status?: TaskStatus;
  team?: string;
  limit?: number;
  offset?: number;
}

interface TaskIdParams {
  id: string;
}

// ---------------------------------------------------------------------------
// JSON schemas
// ---------------------------------------------------------------------------

/** Task ID pattern: lowercase hex chars and hyphens (UUID format). */
const TASK_ID_PATTERN = '^[a-f0-9-]+$';

/** JSON schema for the :id URL parameter. */
export const TASK_ID_PARAM_SCHEMA = {
  params: {
    type: 'object',
    properties: {
      id: { type: 'string', pattern: TASK_ID_PATTERN, maxLength: 128 },
    },
  },
};

/** JSON schema for GET /api/v1/tasks querystring. */
export const GET_TASKS_QUERY_SCHEMA = {
  querystring: {
    type: 'object',
    additionalProperties: false,
    properties: {
      status: {
        type: 'string',
        enum: ['pending', 'running', 'completed', 'failed', 'cancelled'],
      },
      team: {
        type: 'string',
        pattern: '^[a-z0-9][a-z0-9-]*[a-z0-9]$',
        maxLength: 64,
      },
      limit: {
        type: 'integer',
        minimum: 1,
        maximum: MAX_TASKS_LIMIT,
        default: DEFAULT_TASKS_LIMIT,
      },
      offset: {
        type: 'integer',
        minimum: 0,
        default: 0,
      },
    },
  },
};

// ---------------------------------------------------------------------------
// Utility
// ---------------------------------------------------------------------------

/**
 * Builds a TaskWithSubtree from a task and optional subtasks.
 */
export function buildTaskWithSubtree(task: Task, subtasks?: Task[]): TaskWithSubtree {
  return { ...task, subtasks };
}

// ---------------------------------------------------------------------------
// Route handlers
// ---------------------------------------------------------------------------

/**
 * Handler factory for GET /api/v1/tasks.
 * Supports query params: status, team, limit, offset.
 */
export function getTasksHandler(taskStore: TaskStore, _logger: MiddlewareLogger) {
  return async (
    request: FastifyRequest<{ Querystring: GetTasksQuery }>,
    reply: FastifyReply,
  ): Promise<void> => {
    const { status, team } = request.query;
    const limit = request.query.limit ?? DEFAULT_TASKS_LIMIT;
    const offset = request.query.offset ?? 0;

    let tasks: Task[];

    if (team !== undefined) {
      try {
        tasks = await taskStore.listByTeam(team);
      } catch (err) {
        mapDomainError(reply as FastifyReplyShim, err);
        return;
      }
    } else if (status !== undefined) {
      try {
        tasks = await taskStore.listByStatus(status);
      } catch (err) {
        mapDomainError(reply as FastifyReplyShim, err);
        return;
      }
    } else {
      // No filter: list running tasks as practical default.
      try {
        tasks = await taskStore.listByStatus('running');
      } catch {
        tasks = [];
      }
    }

    const total = tasks.length;
    let hasMore = false;
    let paginatedTasks: Task[];

    if (offset < tasks.length) {
      const end = offset + limit;
      if (end >= tasks.length) {
        paginatedTasks = tasks.slice(offset);
      } else {
        hasMore = true;
        paginatedTasks = tasks.slice(offset, end);
      }
    } else {
      paginatedTasks = [];
    }

    const response: TasksResponse = {
      tasks: paginatedTasks,
      total,
      has_more: hasMore,
      limit,
      offset,
    };

    sendJSON(reply as FastifyReplyShim, 200, response);
  };
}

/**
 * Handler factory for GET /api/v1/tasks/:id.
 * Returns task with subtree. Returns 404 if not found.
 */
export function getTaskHandler(taskStore: TaskStore, logger: MiddlewareLogger) {
  return async (
    request: FastifyRequest<{ Params: TaskIdParams }>,
    reply: FastifyReply,
  ): Promise<void> => {
    const { id } = request.params;

    let task: Task;
    try {
      task = await taskStore.get(id);
    } catch (err) {
      mapDomainError(reply as FastifyReplyShim, err);
      return;
    }

    // Fetch subtree; non-fatal if it fails.
    let subtasks: Task[] | undefined;
    try {
      const subtree = await taskStore.getSubtree(id);
      // Filter out root task itself from the subtree.
      subtasks = subtree.filter((t) => t.id !== id);
    } catch (err) {
      logger.warn('failed to get task subtree', err);
    }

    sendJSON(reply as FastifyReplyShim, 200, buildTaskWithSubtree(task, subtasks));
  };
}

/**
 * Handler factory for POST /api/v1/tasks/:id/cancel.
 * Requires Content-Type: application/json for CSRF protection.
 * Returns updated task after cancellation.
 */
export function cancelTaskHandler(
  orch: GoOrchestrator,
  taskStore: TaskStore,
  logger: MiddlewareLogger,
) {
  return async (
    request: FastifyRequest<{ Params: TaskIdParams }>,
    reply: FastifyReply,
  ): Promise<void> => {
    // Enforce JSON Content-Type for CSRF protection (even though body is not read).
    const ct = request.headers['content-type'];
    if (ct !== 'application/json') {
      sendError(
        reply as FastifyReplyShim,
        415,
        'INVALID_CONTENT_TYPE',
        'Content-Type must be application/json',
      );
      return;
    }

    const { id } = request.params;

    try {
      await orch.cancelTask(id);
    } catch (err) {
      logger.error('failed to cancel task', err);
      mapDomainError(reply as FastifyReplyShim, err);
      return;
    }

    // Return updated task.
    try {
      const task = await taskStore.get(id);
      sendJSON(reply as FastifyReplyShim, 200, task);
    } catch (err) {
      mapDomainError(reply as FastifyReplyShim, err);
    }
  };
}

// ---------------------------------------------------------------------------
// Route registration
// ---------------------------------------------------------------------------

/**
 * Registers all task routes on the Fastify instance.
 */
export function registerTaskRoutes(
  fastify: FastifyInstance,
  taskStore: TaskStore,
  orch: GoOrchestrator,
  logger: MiddlewareLogger,
): void {
  fastify.get('/api/v1/tasks', { schema: GET_TASKS_QUERY_SCHEMA }, getTasksHandler(taskStore, logger));
  fastify.get('/api/v1/tasks/:id', { schema: TASK_ID_PARAM_SCHEMA }, getTaskHandler(taskStore, logger));
  fastify.post(
    '/api/v1/tasks/:id/cancel',
    { schema: TASK_ID_PARAM_SCHEMA },
    cancelTaskHandler(orch, taskStore, logger),
  );
}
